pkg/fsutils/fsutils_linux.go

```diff
@@ -0,0 +1,90 @@
+// +build linux
+
+package fsutils
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"syscall"
+	"unsafe"
+)
+
+func locateDummyIfEmpty(path string) (string, error) {
+	children, err := ioutil.ReadDir(path)
+	if err != nil {
+		return "", err
+	}
+	if len(children) != 0 {
+		return "", nil
+	}
+	dummyFile, err := ioutil.TempFile(path, "fsutils-dummy")
+	if err != nil {
+		return "", err
+	}
+	name := dummyFile.Name()
+	if err = dummyFile.Close(); err != nil {
+		return name, err
+	}
+	return name, nil
+}
+
+// SupportsDType returns whether the filesystem mounted on path supports d_type
+func SupportsDType(path string) (bool, error) {
+	// locate dummy so that we have at least one dirent
+	dummy, err := locateDummyIfEmpty(path)
+	if err != nil {
+		return false, err
+	}
+	if dummy != "" {
+		defer os.Remove(dummy)
+	}
+
+	visited := 0
+	supportsDType := true
+	walkFn := func(ent *syscall.Dirent) bool {
+		visited++
+		if ent.Type == syscall.DT_UNKNOWN {
+			supportsDType = false
+			// stop walking
+			return true
+		}
+		// continue walking
+		return false
+	}
+	if err = walkdir(path, walkFn); err != nil {
+		return false, err
+	}
+	if visited == 0 {
+		return false, fmt.Errorf("did not hit any dirent during walking %s", path)
+	}
+	return supportsDType, nil
+}
+
+func walkdir(path string, walkFn func(*syscall.Dirent) bool) error {
```

Comment:
self-nit: `walk` maybe misleading because it is not recursive.
I'll try to change the terminology